### PR TITLE
GH-41502: [Python] Fix reading column index with decimal values

### PR DESCRIPTION
**Pull request by jrversteegh**
Fix for #41502 

Convert pandas "decimal" to "object" in numpy.
* GitHub Issue: #41502

### CODE REVIEW

**Comment by AlenkaF** [comment]
Hi @jrversteegh, thank you for the contribution!

I think this is an elegant solution. Not sure if this was discussed before or not, can't find any similar issue on our issue tracker. I am sure @jorisvandenbossche will know straight away if this change fits or not.

From my side, a test needs to be added in _python/pyarrow/tests/parquet/test_pandas.py_.

**Comment by jorisvandenbossche** [comment]
That indeed looks like a good fix. 

The error itself should already happen with just a roundtrip from pandas->pyarrow->pandas (without parquet), so you can add a test for this in `python/pyarrow/tests/test_pandas.py.`

**Comment by jrversteegh** [comment]
@AlenkaF 
> From my side, a test needs to be added in _python/pyarrow/tests/parquet/test_pandas.py_.

Thanks for that suggestion. I tried, but this issue appears more involved than I expected. It looks like pyarrow expects column names to be strings. If not, it converts them (in turn because the parquet format expects this?).
My fix avoids the exception, but still converts the column names from decimal to string, which is better, but still undesirable. I'll have to see whether there is something I can do about that. To be continued.

**Comment by jrversteegh** [comment]
@AlenkaF  @jorisvandenbossche I've added a test and restored the decimal index from strings. This looks like a bit of a kludge. I think it's because both numpy and pandas don't understand `Decimal`. It's just an object, like any other, so you might expect similar issues for any other object type used as a column index. However, `Decimal`  is not just any object and it does make some sense to use as an index, just like dates, so I feel it's a warranted addition. Let me know what you think.

File: python/pyarrow/tests/test_pandas.py [hunk 1]
```diff
@@ -221,6 +221,16 @@ def test_column_index_names_with_tz(self):
         )
         _check_pandas_roundtrip(df, preserve_index=True)
 
+    def test_column_index_names_with_decimal(self):
```
Comment (AlenkaF) [inline python/pyarrow/tests/test_pandas.py:224]: Could you add info about the GitHub issue this test is added for? Similarly as what you added in  `reconstruct_columns_from_metadata`  or in the test just above.

**Review by AlenkaF — APPROVED** [review]
I think it is an elegant solution still. Thank you for the fix!

File: python/pyarrow/tests/test_pandas.py (hunk 1 above)
Comment (jrversteegh) [reply python/pyarrow/tests/test_pandas.py:224]: @AlenkaF sure, done.

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!

**Pull request merged**